sources/file_table.cpp

```diff
@@ -264,11 +289,49 @@ void FileTable::close(FileBase* fb)
 void FileTable::eject()
 {
     auto num_eject = std::min<size_t>(NUM_EJECT, m_closed_ids.size());
-    for (size_t i = 0; i < num_eject; ++i)
+    /* for (size_t i = 0; i < num_eject; ++i) { */
+    /*         TRACE_LOG("Evicting file with ID=%s from cache", hexify(m_closed_ids[i]).c_str()); */
+    /*         m_files.erase(m_closed_ids[i]); */
+    /* } */
+
     {
-        m_files.erase(m_closed_ids[i]);
-        TRACE_LOG("Evicting file with ID=%s from cache", hexify(m_closed_ids[i]).c_str());
+        std::lock_guard<std::mutex> l(m_closing_lock);
+        for (size_t i = 0; i < num_eject; ++i)
+        {
+            auto id = m_closed_ids[i];
+
+            m_files_to_close.emplace(id,std::move(m_files.at(id)));
+            m_files.erase(id);
+
+            std::function<void()> free_ptr = [id,this]() {
+                FileBase* ptr = nullptr;
+                {
+                    std::lock_guard<std::mutex> l(m_closing_lock);
+                    auto it = m_files_to_close.find(id);
+                    if(it == m_files_to_close.end()) {
+                        // The file was reopened
+                        return;
+                    }
+                    ptr = it->second.release();
```

Comment:
The two lines below may throw exceptions, causing leakage of `ptr`. Instead of raw pointer, it should be a `std::unique_ptr`.